<a href="https://colab.research.google.com/github/santhoshbvsrk/CapstoneProjects/blob/main/DataAnalytics_CDF/InsaidTelecom_CDFCapstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 

*  As the dataset is huge, we'll zip it, upload the zip file to google drive and mount your google drive to colab, unzip it to access data from your colab.


*   Import the drive package from google.colab library to mount your google drive on to google colab.





In [ ]:
from google.colab import drive
drive.mount ('/content/gdrive')



*   Once mounting is completed, mention the path of your zip file which contains the data to read.

In [ ]:
!unzip gdrive/My\ Drive/CDF_Capstone_Project1/events_data.zip



*   Once mounting and unzipping is completed successfully, you should be able to see the zip file contains in your colab folder.


In [ ]:
import pandas as pd 
import numpy as np

In [ ]:
events_data_df=pd.read_csv("events_data.csv")
events_data_df.head()

In [ ]:
events_data_df['state'].value_counts()

Delhi                       751733
Maharashtra                 677105
TamilNadu                   436282
AndhraPradesh               321332
MadhyaPradesh               232690
Rajasthan                   196361
WestBengal                  196161
Karnataka                    99058
UttarPradesh                 94750
Gujarat                      42881
Bihar                        35472
Punjab                       26195
Kerala                       23148
Haryana                      21053
Jharkhand                    18060
Orissa                       15904
Telangana                    12784
Assam                        12021
Chhattisgarh                  9754
Uttaranchal                   7720
JammuandKashmir               5385
Goa                           3250
Nagaland                      2298
Pondicherry                   1741
Mizoram                       1613
AndamanandNicobarIslands      1515
Meghalaya                     1515
HimachalPradesh               1409
Manipur             

Filtering out events_data for only 6 states as per the requirement given.

In [ ]:
events_data_filtered_df= events_data_df[(events_data_df.state == 'WestBengal') | (events_data_df.state == 'Karnataka') | (events_data_df.state == 'Gujarat') | (events_data_df.state == 'Bihar') | (events_data_df.state == 'Punjab') | (events_data_df.state == 'Kerala')]

In [ ]:
events_data_filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 422915 entries, 1 to 3252921
Data columns (total 7 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   event_id   422915 non-null  int64  
 1   device_id  422867 non-null  float64
 2   timestamp  422915 non-null  object 
 3   longitude  422873 non-null  float64
 4   latitude   422873 non-null  float64
 5   city       422915 non-null  object 
 6   state      422915 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 25.8+ MB


Installing mysql connector for python

In [ ]:
pip install mysql-connector-python-rf

     |████████████████████████████████| 11.9MB 12.8MB/s 
  Created wheel for mysql-connector-python-rf: filename=mysql_connector_python_rf-2.2.2-cp37-cp37m-linux_x86_64.whl size=249456 sha256=520c8f18fdbdd6429d5d2ca0eefeaf1d9cbab3399d9f1ef4a9a1a992dbdbbde3
  Stored in directory: /root/.cache/pip/wheels/87/58/fb/d95c84fad7e1bebfed324c13e107ebb08e1997c9226532859a
Successfully built mysql-connector-python-rf


import package to connect to mysql and loading data from tables to dataframes

In [ ]:
import mysql.connector

#this is to connect to mysql database
mydb = mysql.connector.connect(
  host="cpanel.insaid.co",
  user="student",
  password="student",
  database="Capstone1"
)

#declaring a cursor variable of database to hold data of a specific table
mycursor = mydb.cursor()

#mycursor variable holds the entire data from gender_age_train table
mycursor.execute("SELECT * FROM gender_age_train")

#using fetchall function to get data from cursor and populating in another variable
myresult = mycursor.fetchall()

#inserting table data from variable into a dataframe
gender_age_train_df = pd.DataFrame(myresult,columns=['device_id','Gender','Age','Age_Group'])

#mycursor variable holds the entire data from phone_brand_device_model table
mycursor.execute("SELECT * FROM phone_brand_device_model")

#using fetchall function to get data from cursor and populating in another variable, this overwrites data from previous load
myresult = mycursor.fetchall()

#inserting table data from variable into a dataframe
phone_brand_device_model_df = pd.DataFrame(myresult,columns=['device_id','Brand','Model'])

#printing sample data from the dataframes
print("Data from Gender_Age_Train Table")
print("---------------------------------")
print(gender_age_train_df.head())
print("**************************************")
print("\n")
print("Data from Phone_Brand_Device_Model Table")
print("-----------------------------------------")
print(phone_brand_device_model_df.head())
print("**************************************")

Data from Gender_Age_Train Table
---------------------------------
             device_id Gender  Age Age_Group
0 -8076087639492063270      M   35    M32-38
1 -2897161552818060146      M   35    M32-38
2 -8260683887967679142      M   35    M32-38
3 -4938849341048082022      M   30    M29-31
4   245133531816851882      M   30    M29-31
**************************************


Data from Phone_Brand_Device_Model Table
-----------------------------------------
             device_id Brand    Model
0  1877775838486905855  vivo      Y13
1 -3766087376657242966    小米     V183
2 -6238937574958215831  OPPO      R7s
3  8973197758510677470    三星    A368t
4 -2015528097870762664    小米  红米Note2
**************************************


In [ ]:
events_data_filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 422915 entries, 1 to 3252921
Data columns (total 7 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   event_id   422915 non-null  int64  
 1   device_id  422867 non-null  float64
 2   timestamp  422915 non-null  object 
 3   longitude  422873 non-null  float64
 4   latitude   422873 non-null  float64
 5   city       422915 non-null  object 
 6   state      422915 non-null  object 
dtypes: float64(3), int64(1), object(3)
memory usage: 25.8+ MB


In [ ]:
gender_age_train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 74645 entries, 0 to 74644
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  74645 non-null  int64 
 1   Gender     74645 non-null  object
 2   Age        74645 non-null  int64 
 3   Age_Group  74645 non-null  object
dtypes: int64(2), object(2)
memory usage: 2.3+ MB


In [ ]:
phone_brand_device_model_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87726 entries, 0 to 87725
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   device_id  87726 non-null  int64 
 1   Brand      87726 non-null  object
 2   Model      87726 non-null  object
dtypes: int64(1), object(2)
memory usage: 2.0+ MB


In [ ]:
gender_age_train_df['Gender'].value_counts()

M    47904
F    26741
Name: Gender, dtype: int64

No entries in gender_age_train dataframe with null Gender

In [ ]:
phone_brand_device_model_df['Brand'].value_counts(ascending=False)

小米      21645
三星      19211
华为      15181
vivo     7596
OPPO     6717
        ...  
E人E本        1
丰米          1
西门子         1
世纪星         1
恒宇丰         1
Name: Brand, Length: 116, dtype: int64

No records in phone_brand_device_model dataframe with null Brand

Merging Dataframes on device_id to create a single dataframe which contains customer age, gender, mobile brand and model

In [ ]:
device_data_df=pd.merge(events_data_filtered_df,gender_age_train_df,on='device_id',how='inner') #performing inner join to fetch only matching device_id records
device_data_df.fillna(0,inplace=True) #replacing null values in the columns with 0
device_data_df

event_id     device_id            timestamp  ...  Gender  Age Age_Group
0      1836426  3.872522e+18  2016-05-01 20:29:14  ...       M   20      M22-
1       225132  3.872522e+18  2016-05-01 12:52:44  ...       M   20      M22-
2       815723  3.872522e+18  2016-05-01 12:28:56  ...       M   20      M22-
3      1300157  3.872522e+18  2016-05-04 23:24:57  ...       M   20      M22-
4       233682  3.872522e+18  2016-05-04 19:09:26  ...       M   20      M22-
...        ...           ...                  ...  ...     ...  ...       ...
2636   1530731 -2.288468e+17  2016-05-07 10:35:00  ...       F   63      F43+
2637   2967042 -2.288468e+17  2016-05-01 08:08:46  ...       F   63      F43+
2638   1334106 -2.734689e+18  2016-05-05 10:00:05  ...       M   27    M27-28
2639   1776773  3.711362e+18  2016-05-02 15:54:13  ...       M   31    M29-31
2640   1608869  3.501847e+18  2016-05-04 13:27:20  ...       M   28    M27-28

[2641 rows x 10 columns]

In [ ]:
device_data_df=pd.merge(device_data_df,phone_brand_device_model_df,on='device_id',how='inner')
device_data_df.fillna(0,inplace=True)
device_data_df

event_id     device_id            timestamp  ...  Age_Group  Brand      Model
0      1836426  3.872522e+18  2016-05-01 20:29:14  ...       M22-     酷派       8702
1       225132  3.872522e+18  2016-05-01 12:52:44  ...       M22-     酷派       8702
2       815723  3.872522e+18  2016-05-01 12:28:56  ...       M22-     酷派       8702
3      1300157  3.872522e+18  2016-05-04 23:24:57  ...       M22-     酷派       8702
4       233682  3.872522e+18  2016-05-04 19:09:26  ...       M22-     酷派       8702
...        ...           ...                  ...  ...        ...    ...        ...
2636   1530731 -2.288468e+17  2016-05-07 10:35:00  ...       F43+     魅族        MX3
2637   2967042 -2.288468e+17  2016-05-01 08:08:46  ...       F43+     魅族        MX3
2638   1334106 -2.734689e+18  2016-05-05 10:00:05  ...     M27-28     华为    荣耀3X畅玩版
2639   1776773  3.711362e+18  2016-05-02 15:54:13  ...     M29-31     三星  Galaxy S6
2640   1608869  3.501847e+18  2016-05-04 13:27:20  ...     M27-28     小米         E5

[2641 rows x 12 columns]

In [ ]:
device_data_df['device_id'].value_counts()

 1.274328e+17    285
-1.116277e+18    224
 7.636862e+16    151
-2.876037e+15     76
 4.566137e+14     75
                ... 
 3.711362e+18      1
 3.299619e+17      1
-2.056028e+17      1
-9.869566e+15      1
 1.512378e+18      1
Name: device_id, Length: 120, dtype: int64

1.   Only 2641 records have matching device_id from all the 3 dataframes.
2.   Of 2641 matching records there are 120 unique DeviceIDs which mean that there are multiple events for few Devices.


In [ ]:
device_data_df['Model'].value_counts(ascending=True)

E5              1
T1              1
Z9              1
超级手机1           1
红米note增强版       2
             ... 
荣耀畅玩4         115
MI 3          137
Galaxy Win    151
红米Note2       224
红米note        306
Name: Model, Length: 85, dtype: int64

In [ ]:
device_data_df['Brand'].value_counts(ascending=True)

锤子        1
酷比魔方      8
努比亚       9
亿通       17
一加       35
HTC      57
乐视       72
魅族       94
OPPO     95
酷派      161
vivo    205
华为      369
三星      577
小米      941
Name: Brand, dtype: int64

Upon merging all the 3 dataframes we've data for around 85 models from 14 Brands.

In [ ]:
device_data_df['state'].value_counts()

WestBengal    959
Gujarat       707
Karnataka     572
Bihar         300
Kerala         71
Punjab         32
Name: state, dtype: int64

# Observations


1.   There are 
2.   List item

